### ISSUE TITLE
Failing to set wallpaper on start up.

### ISSUE THREAD

**Issue by nervocalm** [Blacksuan19/plasma-wallpaper-wallhaven-reborn#6]
Hi guys!
I don't entirely sure this is a "plugin problem" but it happens to me at least that when booting up into Plasma, the plugin request the wallpaper before the connection is established, and then fail. Is not the end of the world, right-click on desktop and select refresh and it loads without a problem. But may be it can be improved and check for the availability of the URL before making the request? and in case that is not available, then it may retry a few times with a few seconds delay before  giving up.
In the logs is easy to see than in the line 1 make the request and in the line 12 gives up, and  in the middle is managing to connect to the wifi. It all happens in the same second.
Curiously (and off topic), in the line 13-14 the same happens with bing "picture of the day" that I have set in the other monitor but that one seems to rebound after because it refreshes the wallpaper after without manual intervention.

```
1. Sep 11 10:27:44 smartg12 plasmashell[941]: qml: using url: https://wallhaven.cc/api/v1/search?categories=111&purity=111&sorting=date_added&page=1&atleast=1920x1080&ratios=16x9&q=landscape%20mountains
2. Sep 11 10:27:44 smartg12 wpa_supplicant[680]: wlo1: WPA: Key negotiation completed with 6c:73:7e:91:d4:77 [PTK=CCMP GTK=CCMP]
3. Sep 11 10:27:44 smartg12 wpa_supplicant[680]: wlo1: CTRL-EVENT-CONNECTED - Connection to 6c:73:7e:91:d4:77 completed [id=0 id_str=]
4. Sep 11 10:27:44 smartg12 NetworkManager[602]: <info>  [1726061264.5424] device (wlo1): supplicant interface state: 4way_handshake -> completed
5. Sep 11 10:27:44 smartg12 wpa_supplicant[680]: wlo1: CTRL-EVENT-SIGNAL-CHANGE above=1 signal=-44 noise=9999 txrate=26000
6. Sep 11 10:27:44 smartg12 NetworkManager[602]: <info>  [1726061264.5425] device (wlo1): Activation: (wifi) Stage 2 of 5 (Device Configure) successful. Connected to wireless network "Fibertel WiFi u3o 2.4GHz"
7. Sep 11 10:27:44 smartg12 NetworkManager[602]: <info>  [1726061264.5425] device (p2p-dev-wlo1): supplicant management interface state: 4way_handshake -> completed
8. Sep 11 10:27:44 smartg12 NetworkManager[602]: <info>  [1726061264.5429] device (wlo1): state change: config -> ip-config (reason 'none', sys-iface-state: 'managed')
9. Sep 11 10:27:44 smartg12 NetworkManager[602]: <info>  [1726061264.5441] dhcp4 (wlo1): activation: beginning transaction (timeout in 45 seconds)
10. Sep 11 10:27:44 smartg12 NetworkManager[602]: <info>  [1726061264.6103] dhcp4 (wlo1): state changed new lease, address=192.168.1.12, acd pending
11. Sep 11 10:27:44 smartg12 remmina[1177]: [X2GO] X2Go plugin loaded.
12. Sep 11 10:27:44 smartg12 plasmashell[941]: qml: getImageData Error:failed to send request
13. Sep 11 10:27:44 smartg12 plasmashell[941]: kf.kio.core: mimeType() not emitted when sending first data!; job URL = QUrl("https://www.bing.com/HPImageArchive.aspx?format=js&idx=0&n=1") data size = 0
14. Sep 11 10:27:44 smartg12 plasmashell[941]: kde.wallpapers.potd: "bing" with arguments QList(QVariant(int, 1920), QVariant(int, 1080)) failed to fetch the remote wallpaper. Please check your Internet connection>
15. Sep 11 10:27:44 smartg12 kded6[917]: Registering ":1.49/StatusNotifierItem" to system tray
16. Sep 11 10:27:44 smartg12 remmina[1177]: kf.windowsystem: Cannot use KWindowSystem without a QGuiApplication
17. Sep 11 10:27:44 smartg12 NetworkManager[602]: <info>  [1726061264.8067] dhcp4 (wlo1): state changed new lease, address=192.168.1.12
18. Sep 11 10:27:44 smartg12 NetworkManager[602]: <info>  [1726061264.8077] policy: set 'Fibertel WiFi u3o 2.4GHz' (wlo1) as default for IPv4 routing and DNS
19. Sep 11 10:27:44 smartg12 dbus-broker-launch[592]: Activation request for 'org.freedesktop.resolve1' failed: The systemd unit 'dbus-org.freedesktop.resolve1.service' could not be found.
20. Sep 11 10:27:44 smartg12 pamac-tray-plas[1192]: updates_checker.vala:100: 0 updates found
21. Sep 11 10:27:44 smartg12 NetworkManager[602]: <info>  [1726061264.8191] device (wlo1): state change: ip-config -> ip-check (reason 'none', sys-iface-state: 'managed')
22. Sep 11 10:27:44 smartg12 systemd[1]: Starting Network Manager Script Dispatcher Service...
23. Sep 11 10:27:44 smartg12 systemd[1]: Started Network Manager Script Dispatcher Service.
24. Sep 11 10:27:44 smartg12 NetworkManager[602]: <info>  [1726061264.8595] device (wlo1): state change: ip-check -> secondaries (reason 'none', sys-iface-state: 'managed')
25. Sep 11 10:27:44 smartg12 NetworkManager[602]: <info>  [1726061264.8600] device (wlo1): state change: secondaries -> activated (reason 'none', sys-iface-state: 'managed')
26. Sep 11 10:27:44 smartg12 NetworkManager[602]: <info>  [1726061264.8605] manager: NetworkManager state is now CONNECTED_SITE
27. Sep 11 10:27:44 smartg12 NetworkManager[602]: <info>  [1726061264.8615] device (wlo1): Activation: successful, device activated.
28. Sep 11 10:27:44 smartg12 systemd[783]: Starting Accessibility services bus...
```

I don't know if I can modify this in plasma configuration (some sort of load order or something).

Thanks!

**Comment by Blacksuan19** [comment]
I can't seem to reproduce this, when I first login, the screen is black for few seconds then a wallpaper is loaded as expected with no need to manually trigger a refresh first, could be because your internet is not connected yet, I remember having that issue on plasma 5 with the original plugin, anyway, a retry request feature is a good idea so I will add it shortly

**Comment by Blacksuan19** [comment]
implemented on the latest commit, test and let me know if it works.

**Comment by nervocalm** [comment]
Wow, that was fast!
I just did a boot with my current config and still got black screen.
Then I thought that may be it has something to do with being set on the external monitor, so I configure wallhaven in both screens and then I did another boot. And then both screens were black too but in the second boot I get another set of errors (very similar to the previous issue you had posted, may be I can use what you did there to resolve it).
I filter a little bit the log because was very long. At 14:00 the fist boot and at 14:11 the second. The query in the integrated screen is "sunset sky" and in the external monitor "landscape mountains".

```
Sep 11 14:00:25 smartg12 plasmashell[911]: qml: using url: https://wallhaven.cc/api/v1/search?categories=111&purity=111&sorting=date_added&page=1&atleast=1920x1080&ratios=16x9&q=landscape%20mountains
Sep 11 14:00:26 smartg12 plasmashell[911]: file:///home/stomasin/.local/share/plasma/wallpapers/com.plasma.wallpaper.wallhaven/contents/ui/main.qml:140: Error: Invalid write to global property "msg"
Sep 11 14:11:35 smartg12 plasmashell[911]: kf.coreaddons: "Could not find plugin com.plasma.wallpaper.wallhaven"
Sep 11 14:11:35 smartg12 plasmashell[911]: file:///home/stomasin/.local/share/plasma/wallpapers/com.plasma.wallpaper.wallhaven/contents/ui/config.qml:63:9: Unable to assign [undefined] to QUrl
Sep 11 14:11:39 smartg12 plasmashell[911]: kf.coreaddons: "Could not find plugin com.plasma.wallpaper.wallhaven"
Sep 11 14:12:08 smartg12 plasmashell[911]: qml: using url: https://wallhaven.cc/api/v1/search?categories=111&purity=100&sorting=date_added&page=1&atleast=1920x1080&ratios=16x9&q=sunset%20sky
Sep 11 14:12:08 smartg12 plasmashell[911]: file:///home/stomasin/.local/share/plasma/wallpapers/com.plasma.wallpaper.wallhaven/contents/ui/config.qml:32: TypeError: Cannot read property 'configuration' of null
Sep 11 14:12:08 smartg12 plasmashell[911]: file:///home/stomasin/.local/share/plasma/wallpapers/com.plasma.wallpaper.wallhaven/contents/ui/config.qml:63: TypeError: Cannot read property 'currentWallpaperThumbnail' of null
Sep 11 14:12:08 smartg12 plasmashell[911]: file:///home/stomasin/.local/share/plasma/wallpapers/com.plasma.wallpaper.wallhaven/contents/ui/config.qml:64: TypeError: Cannot read property 'currentWallpaperThumbnail' of null
Sep 11 14:13:39 smartg12 plasmashell[914]: qml: using url: https://wallhaven.cc/api/v1/search?categories=111&purity=100&sorting=date_added&page=1&atleast=1920x1080&ratios=16x9&q=sunset%20sky
Sep 11 14:13:40 smartg12 plasmashell[914]: qml: using url: https://wallhaven.cc/api/v1/search?categories=111&purity=111&sorting=date_added&page=1&atleast=1920x1080&ratios=16x9&q=landscape%20mountains
Sep 11 14:13:40 smartg12 plasmashell[914]: file:///home/stomasin/.local/share/plasma/wallpapers/com.plasma.wallpaper.wallhaven/contents/ui/main.qml:140: Error: Invalid write to global property "msg"
Sep 11 14:13:40 smartg12 plasmashell[914]: file:///home/stomasin/.local/share/plasma/wallpapers/com.plasma.wallpaper.wallhaven/contents/ui/main.qml:140: Error: Invalid write to global property "msg"

```

**Comment by Blacksuan19** [comment]
I think wallpaper is set per monitor so you will have to manually configure the plugin for each monitor, I have pushed a commit to address the `msg` errors, everything else is normal initially before there is a wallpaper set

**Comment by nervocalm** [comment]
yes, is set by monitor and I did that manually. But I mention that because I got more error messages from the main monitor than from the external one.
Did you see the difference??
In the external monitor got two lines, but when I set it also in the main more messages arise (like the ones with TypeError: Cannot read property). Those are similar to the ones you have posted before in the issue with "help needed" but you already close it. How do you resolve that? may be it can help.
I'm going to test the new build an let you know.

**Comment by nervocalm** [comment]
Great!
Now it works. In both monitors...
I rebooted 3 times (to be sure) and got images every time.
I have it configured to cycle each 60 minutes so I got the same images, but now starts black for a few seconds, and then loads them up. I can even see in the notification, the failed attempt (when the connection is not ready yet) and the successful ones.

![wallhaven](https://github.com/user-attachments/assets/6d0fcf1d-bae7-4c5f-b000-6a72fc5fe86e)

Thanks man!

**Comment by nervocalm** [comment]
One thing I noticed is that is always load the same images initially.
And if a refresh manually load a new one but is always in the same order.
And on reboot, I came back to the first one on the previous session.
For example:
Reboot - image1
refresh - image2
regresh - image3

Reboot - image1
refresh -image2
...
This are my settings:
![wallhaven settings](https://github.com/user-attachments/assets/a65b66e2-81a9-4189-b58c-52bcfcde050f)

I have a sense that when I first try this it were more random.
Am I wrong?
Need I to reset some cache or something?
Probably it has nothing to do with the plugin, just asking in case you know.

Thanks!
